lib/linklist.c

```diff
@@ -331,7 +331,7 @@ void list_delete_all_node(struct list *list)
 		next = node->next;
 		if (*list->del)
 			(*list->del)(node->data);
-		listnode_free(list, node);
+		list_delete_node(list, node);
```

Comment:
```
void list_delete_node(struct list *list, struct listnode *node)
{
	frrtrace(2, frr_libfrr, list_delete_node, list, node);

	if (node->prev)
		node->prev->next = node->next;
	else
		list->head = node->next;
	if (node->next)
		node->next->prev = node->prev;
	else
		list->tail = node->prev;
	list->count--;
	listnode_free(list, node);
}
```

`list_delete_node()` updates the link-list and free the node with listnode_free(). This is correct but, since all the nodes are deleted, we do not need to do that. listnode_free() is sufficient

Reply:
> `list_delete_node()` updates the link-list and free the node with listnode_free(). This is correct but, since all the nodes are deleted, we do not need to do that. listnode_free() is sufficient

I don't think so. Refer to the backtrace above, ` (*list->del)()=>...=>isis_level2_adj_up()`. If the memory of each node in the list is not unlinked before being released, the operation to delete node data `(*list->del)()` should not be allowed to continue traversing this linked list, as it would otherwise access released node memory. However, the restriction on allowing `(*list->del)()` to traverse the list is inherently unreasonable.

```
bool isis_level2_adj_up(struct isis_area *area)
{
	struct listnode *node, *cnode;
	struct isis_circuit *circuit;
	struct list *adjdb;
	struct isis_adjacency *adj;

	if (area->is_type == IS_LEVEL_1)
		return false;

	for (ALL_LIST_ELEMENTS_RO(area->circuit_list, cnode, circuit)) {
		if (circuit->circ_type == CIRCUIT_T_BROADCAST) {
			adjdb = circuit->u.bc.adjdb[1];
			if (!adjdb || !adjdb->count)
				continue;

			for (ALL_LIST_ELEMENTS_RO(adjdb, node, adj)) {    //here heap-after-free
				if (adj->level != ISIS_ADJ_LEVEL1
				    && adj->adj_state == ISIS_ADJ_UP)
					return true;
			}
		} else if (circuit->circ_type == CIRCUIT_T_P2P
			   && circuit->u.p2p.neighbor) {
			adj = circuit->u.p2p.neighbor;
			if (adj->level != ISIS_ADJ_LEVEL1
			    && adj->adj_state == ISIS_ADJ_UP)
				return true;
		}
	}

	return false;
}
```